pkg/auth/artifact_verifier.go

```diff
@@ -0,0 +1,162 @@
+// was created by a trusted entity.
+type ArtifactVerifier interface {
+	VerifyHoistArtifact(localCopy *os.File, artifactLocation string) error
+}
+
+type noopVerifier struct{}
+
+func (n *noopVerifier) VerifyHoistArtifact(_ *os.File, _ string) error {
+	return nil
+}
+
+func NoopVerifier() ArtifactVerifier {
+	return &noopVerifier{}
+}
+
+// BuildManifestVerifier ensures that the given launchable's location
+// field is matched with a corresponding manifest certifying the validity
+// of the build. The manifest is a YAML file containing a single key "artifact_sha".
+// That key represents the hex digest of the artifact tar itself. The manifest is
+// signed by the build system and has a corresponding URL for downloading the signature.
+//
+// The manifest and signature files should match the following convention if the launchable
+// location is a full path to a file:
+//
+// If the artifact is located here:
+// https://foo.bar.baz/artifacts/myapp_abc123.tar.gz
+//
+// Then its build manifest is located here:
+// https://foo.bar.baz/artifacts/myapp_abc123.tar.gz.manifest
+//
+// artifact_sha: abc23456
+//
+// And its signature file is located here:
+// https://foo.bar.baz/artifacts/myapp_abc123.tar.gz.manifest.sig
+type BuildManifestVerifier struct {
+	keyring openpgp.KeyRing
+	fetcher uri.Fetcher
+	logger  *logging.Logger
+}
+
+func NewBuildArtifactVerifier(keyringPath string, fetcher uri.Fetcher, logger *logging.Logger) (*BuildManifestVerifier, error) {
+	keyring, err := LoadKeyring(keyringPath)
+	if err != nil {
+		return nil, util.Errorf("Could not load artifact verification keyring from %v: %v", keyringPath, err)
+	}
+	return &BuildManifestVerifier{
+		keyring: keyring,
+		fetcher: fetcher,
+		logger:  logger,
+	}, nil
+}
+
+// Returns an error if the stanza's artifact is not signed appropriately. Note that this
+// implementation does not use the pod manifest digest location options.
+func (b *BuildManifestVerifier) VerifyHoistArtifact(localCopy *os.File, artifactLocation string) error {
+	u, err := url.Parse(artifactLocation)
+	if err != nil {
+		return err
+	}
+	switch u.Scheme {
+	default:
+		return fmt.Errorf("%v does not have a recognized scheme, cannot verify manifest or signature", artifactLocation)
+	case "http", "https", "file":
+		dir, err := ioutil.TempDir("", "artifact_verification")
+		if err != nil {
+			return fmt.Errorf("Could not create temporary directory for manifest file: %v", err)
+		}
+		defer os.RemoveAll(dir)
+
+		manifestSrc := fmt.Sprintf("%v.manifest", artifactLocation)
+		manifestDst := filepath.Join(dir, "manifest")
+		err = b.fetcher.CopyLocal(manifestSrc, manifestDst)
+		if err != nil {
+			return fmt.Errorf("Could not download artifact manifest for %v: %v", artifactLocation, err)
+		}
+
+		signatureSrc := fmt.Sprintf("%v.sig", manifestSrc)
+		signatureDst := filepath.Join(dir, "signature")
+		if err = b.fetcher.CopyLocal(signatureSrc, signatureDst); err != nil {
```

Comment:
nit: you're using different `err` checking syntax between the two `CopyLocal` calls